zipline/src/nativeMain/kotlin/app/cash/zipline/InboundCallChannel.kt

```diff
@@ -0,0 +1,161 @@
+
+    val globalThis = JS_GetGlobalObject(context)
+    val inboundChannel = JS_GetPropertyStr(context, globalThis, inboundChannelName)
+    val property = JS_NewAtom(context, "invoke")
+    val arg0 = JS_NewString(context, instanceName)
+    val arg1 = JS_NewString(context, funName)
+    val arg2 = with(quickJs) { encodedArguments.toJsValue() }
+
+    val jsResult = memScoped {
+      val args = allocArray<JSValue>(3)
+      args[0] = arg0
+      args[1] = arg1
+      args[2] = arg2
+
+      JS_Invoke(context, inboundChannel, property, 3, args)
+    }
+    val kotlinResult = with(quickJs) { jsResult.toKotlinInstanceOrNull() } as Array<String>
+
+    JS_FreeValue(context, jsResult)
+    JS_FreeValue(context, arg2)
+    JS_FreeValue(context, arg1)
+    JS_FreeValue(context, arg0)
+    JS_FreeAtom(context, property)
+    JS_FreeValue(context, inboundChannel)
+    JS_FreeValue(context, globalThis)
+
+    return kotlinResult
+  }
+
+  override fun invokeSuspending(
+    instanceName: String,
+    funName: String,
+    encodedArguments: Array<String>,
+    callbackName: String,
+  ) {
+    quickJs.checkNotClosed()
+
+    val globalThis = JS_GetGlobalObject(context)
+    val inboundChannel = JS_GetPropertyStr(context, globalThis, inboundChannelName)
+    val property = JS_NewAtom(context, "invokeSuspending")
+    val arg0 = JS_NewString(context, instanceName)
+    val arg1 = JS_NewString(context, funName)
+    val arg2 = with(quickJs) { encodedArguments.toJsValue() }
+    val arg3 = JS_NewString(context, callbackName)
+
+    val jsResult = memScoped {
+      val args = allocArray<JSValue>(4)
+      args[0] = arg0
+      args[1] = arg1
+      args[2] = arg2
+      args[3] = arg3
+
+      JS_Invoke(context, inboundChannel, property, 4, args)
+    }
+
+    JS_FreeValue(context, jsResult)
+    JS_FreeValue(context, arg3)
+    JS_FreeValue(context, arg2)
+    JS_FreeValue(context, arg1)
+    JS_FreeValue(context, arg0)
+    JS_FreeAtom(context, property)
+    JS_FreeValue(context, inboundChannel)
+    JS_FreeValue(context, globalThis)
+  }
+
+  override fun disconnect(instanceName: String): Boolean {
+    quickJs.checkNotClosed()
+
+    val globalThis = JS_GetGlobalObject(context)
+    val inboundChannel = JS_GetPropertyStr(context, globalThis, inboundChannelName)
+    val property = JS_NewAtom(context, "disconnect")
+    val arg0 = JS_NewString(context, instanceName)
+
+    val jsResult = memScoped {
+      val args = allocArray<JSValue>(1)
+      args[0] = arg0
+
+      JS_Invoke(context, inboundChannel, property, 1, args)
+    }
+    val kotlinResult = with(quickJs) { jsResult.toKotlinInstanceOrNull() } as Boolean
+
+    JS_FreeValue(context, jsResult)
+    JS_FreeAtom(context, property)
+    JS_FreeValue(context, inboundChannel)
+    JS_FreeValue(context, globalThis)
+
+    return kotlinResult
+  }
+
+  private inline operator fun CArrayPointer<JSValue>.set(index: Int, value: CValue<JSValue>) {
+    val target = this[index]
+    value.useContents {
```

Comment:
I have no idea, do you need to zero fill this at all?

Reply:
No, but this code goes away in the next PR where I discovered how to do a normal struct copy into a pointer offset rather than this piecewise copying.